### ISSUE TITLE
constructor injection vs resolver for DI

### ISSUE THREAD

**Issue by PJB3005** [space-wizards/RobustToolbox#221]
This is an important design choice I'm not sure about yet.

With dependency injection, there's a couple ways to get the dependencies. There's **constructor injection**:

```cs
public class Foo
{
    private readonly ILogger Logger;
    Foo(ILogger logger)
    {
        Logger = logger;
    }
}
```

And there's **resolvers**:

```cs
public class Foo
{
    private readonly ILogger Logger;
    Foo()
    {
        Logger = IoCManager.Resolve<ILogger>();
    }
}
```

Both of these approaches have their up and downsides, and I have a hard time deciding so I'm making this discussion.

Advantages of **constructor injection** (to **resolvers**):
* Allows parallelized unit tests.
* Makes dependency graphs easier to track (read: easier to see what goes where)
* Infinitely easier to manage state properly. You can actually have separate containers.

Disadvantages:
* Circular references are impossible, and require abstraction into even smaller components, which often just makes the code hard to follow so you have your idealistic DI pattern.
* **Confusing**
* Harder to manage. Constructors on MANY classes will now need this.
* Memory inefficient because even uncommon services need to have references kept.

Personally I feel like **resolvers** are the better option because they're much easier to work with. **Issue** is that if we need unit tests to be parallel (and we probably do want that) or we need to be able to keep multiple containers we cannot go this route and we're forced to **constructor injection**.

**Comment by ZoldorfTheWizard** [comment]
With the amount of stuff we would need to pass in a constructor and the thought of that refactor I vote for resolvers right off the start. 

Aren't unit tests currently running parallel? what is preventing them from doing so? is IoC only serving one at a time?

Does this also apply to the container issue you mentioned? What prevents multiple containers?

**Comment by PJB3005** [comment]
You can't have multiple containers when IoC is static, because I believe unit tests simply run on different threads, so the unit tests would modify the static variable at the same time and break everything.

**Comment by YotaXP** [comment]
Could modify IoC to use [`ThreadLocal<T>`](https://msdn.microsoft.com/en-us/library/dd642243(v=vs.110).aspx) so that each thread has their own version.

**Comment by PJB3005** [comment]
That would ruin any chance of multithreading later down the line.

**Comment by PJB3005** [comment]
@volundr- I'd like some input from you too on this issue.

**Comment by psykzz** [comment]
Is it possible to remove them from constructors and maybe us annotations to inject the deps?

Either either way, i'd like to remove the IoC calls but also don't want to see a bunch of stuff added to a constructor, that should be abstracted away from me.

**Comment by psykzz** [comment]
Something i've seen is like this here - https://msdn.microsoft.com/en-us/library/ff660864(v=pandp.20).aspx

Example
```
public class ProductService
{
  private SupplierData supplier;

  [Dependency]
  public SupplierData SupplierDetails 
  {
    get { return supplier; }
    set { supplier = value; }
  }
} 
```

**Comment by PJB3005** [comment]
@psykzz sounds good.

**Comment by PJB3005** [comment]
Could probably do some code generation magic to speed that up too. I'm down.

**Comment by PJB3005** [comment]
Right now I'm leaning towards static resolvers for convenience and property injection as the preferred option.

**Comment by PJB3005** [comment]
I think we've decided.
